cadence/freshmint-encoding/FreshmintEncoding.cdc

```diff
@@ -0,0 +1,105 @@
+/// FreshmintEncoding is a set of utilities for encoding Cadence values as byte arrays.
+///
+/// Variable-legnth values (String, Int, UInt) include a fixed-size length prefix to
+/// prevent distinct sets of values from encoding to the same byte sequence when concatenated.
+///
+/// For example, for a structure with two fields:
+///
+///  - foo: String
+///  - bar: String
+///
+/// Without a length prefix, these instances would produce the same encoding:
+///
+///  - Instance A: { foo: "foo", bar: "bar" }
+///  - Instance b: { foo: "foob", bar: "ar" }
+///
+/// when using an encoding function like this:
+/// 
+/// let encoding = encodeString(foo).concat(encodeString(bar))
+///
+pub contract FreshmintEncoding {
+
+    pub fun encodeUInt(_ value: UInt): [UInt8] {
+        let bytes = value.toBigEndianBytes()
+        let length = UInt16(bytes.length)
+
+        // Prefix byte encoding with a fixed-size encoding of its length
+        return length.toBigEndianBytes().concat(bytes)
+    }
+
+    pub fun encodeUInt8(_ value: UInt8): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeUInt16(_ value: UInt16): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeUInt32(_ value: UInt32): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeUInt64(_ value: UInt64): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeUInt128(_ value: UInt128): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeUInt256(_ value: UInt256): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeInt(_ value: Int): [UInt8] {
```

Comment:
For all the signed cases below (`Int`, `Int8`..), I'm wondering how is the sign encoded.

Reply:
Cadence uses two's complement: https://github.com/onflow/cadence/blob/3db71b8364aee60a83dd53d8a99e935a0e5c8b78/runtime/interpreter/big.go#L27-L58